proposals/0110-exponential-fee-for-write-lock-accounts.md

```diff
@@ -0,0 +1,138 @@
+Exponential Moving Average (EMA) of contentious accounts' CU utilization
+per block and exponentially increasing the cost to write-lock these accounts
+if their utilization remains high.
+
+## Motivation
+
+The motivation behind this feature is to introduce economic back pressure,
+dissuading DeFi spammers from overwhelming the network. DeFi spam, defined as
+opportunistic trades with a positive return on investment, currently occupies
+almost half of the CUs in a block as failed transactions. Economic back
+pressure aims to create a deterrent for such spam activities, ensuring network
+resources are efficiently utilized and preventing continuous block congestion
+caused by failed DeFi spam transactions.
+
+## Alternatives Considered
+
+While the priority fee serves to mitigate low-cost spams by decreasing the
+likelihood of less prioritized transactions being included, it cannot entirely
+eliminate the inclusion of spam transactions in a block. As long as there
+remains a chance, no matter how small, to inexpensively include transactions,
+the incentive for spamming persists. The proposed feature recognizes that the
+current mechanisms have limitations in fully addressing the spam issue, and
+thus, seeks to introduce a more robust solution to discourage opportunistic
+trades and ensure a more secure and efficient network environment.
+
+## New Terminology
+
+- *compute-unit utilization*: denominated in `cu`, it represents total
+ compute-units applied to a given resource.
+- *cost rate*: denominated in `lamport/cu`, it represents the cost per
+compute-unit at a given condition.
+- *compute unit pricer*: a componenet tracks Exponential Moving Average of
+*compute-unit utilization*, applies a pricing algorithm to provide current
+*cost rate*.
+- *write lock fee*: denominated in `lamport`, it is fee dedicated for write
+lock an account, calculated as `compute-unit-pricer.cost-rate() * transaction.requested_cu()`.
+
+## Design Highlights
+
+- Account Association with Compute Unit Pricer:
+  - Accounts are associated with a *compute unit pricer*, and the *runtime*
+  maintains an LRU cache of actively contentious accounts' public keys and
+  their *compute unit pricers*.
+  - Alternatively, each account can have its *compute unit pricer* stored
+  onchain, which would require modifying accounts.
+- Compute Unit Pricer:
+  - Tracks an account's EMA *compute-unit utilization*, updated after the
+  current bank is frozen.
+  - Provides the current *cost rate* when queried.
+- EMA of Compute-Unit Utilization:
+  - Uses 150 slots for EMA calculation.
+  - EMA Alpha, representing the degree of weighting decrease, is calculated as
+  `alpha = 2 / (N+1)`.
+- Pricing Algorithm:
+  - Adjusts write-lock *cost rate* based on an account's EMA *compute-unit
+  utilization*. Initial write-lock cost rate is `1000 lamport/CU`.
+  - For each block, if an account's EMA *compute-unit utilization* is more than
```

Comment:
why a discrete change at 50%? why not `increase_pct = f(utilization)` where f(0%) = -max, f(100%) = +max (perhaps max = 1%) with a continuous function (e.g. linear)?

Reply:
and if it's gonna be discrete, why not e.g. 33%?

Reply:
the shape of the controller deserves some study. understood there is a desire for tx senders to know what the max they might pay is. price as an increasing function of utilization (which all these proposals provide) is already much better than status quo, but a significantly more aggressive increase would help with discrete opportunities like NFT mints and one-time arbs

Reply:
Yea. I agree. I think it's worth picking something reasonable and then adjusting it in the future.  1% per block increase puts the worst case on a tx to 4.4x

Reply:
Discrete makes it easier to manage a cache. Accounts that cross 6m evict the cheapest from the cache.

Reply:
Related to @eugene-chen's comment: the choice of 50% implies some target ("utilization <= 50% is fine; over 50% is undesirable"). Since the EMA is tracked anyway, could you have `inc_pct = f(utilization)` be something like `constant * (utilization - target)` ? 

Since the increase/decrease is in percentage, implicitly the calculation is being done in "logarithmic space" so `f(utilization) = exp(constant * (utilization - target) )` is more appropriate. (We have some work that suggests that the current price should probably appear in the exponent as well: see appendix c of [this paper](https://arxiv.org/pdf/2208.07919.pdf).)

Reply:
discrete is a better devX for clients to figure out pricing